File: java/242.valid-anagram.java
/*
 * @lc app=leetcode id=242 lang=java
 *
 * [242] Valid Anagram
 */
class Solution {
    public boolean isAnagram(String s, String t) {
        if (s == null || t == null || s.length() != t.length()) return false;
        int[] freq = new int[26];
        for (int i = 0; i < s.length(); ++i) freq[s.charAt(i)-'a']++;
        for (int i = 0; i < t.length(); ++i) {
            if (--freq[t.charAt(i)-'a'] < 0) return false;
        }
        return true;
    }
}

File: go/242.valid-anagram.go
/*
 * @lc app=leetcode id=242 lang=golang
 *
 * [242] Valid Anagram
 */

// @lc code=start
func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	freq := make(map[rune]int)
	rt := []rune(t)
	for _, v := range rt {
		if f, b := freq[v]; b {
			freq[v] = f + 1
		} else {
			freq[v] = 1
		}
	}
	rs := []rune(s)
	for _, v := range rs {
		if f, b := freq[v]; b {
			if f == 0 {
				return false
			}
			freq[v] = f - 1
		} else {
			return false
		}
	}
	return true
}

// @lc code=end

